core/trino-main/src/main/java/io/trino/sql/dialect/trino/operation/Join.java

```diff
@@ -0,0 +1,270 @@
+    private final Map<AttributeKey, Object> attributes;
+
+    public Join(
+            String resultName,
+            Value left,
+            Value right,
+            Block leftCriteriaSelector,
+            Block rightCriteriaSelector,
+            Block filter,
+            Block leftHashSelector,
+            Block rightHashSelector,
+            Block leftOutputSelector,
+            Block rightOutputSelector,
+            Block dynamicFilterTargetSelector,
+            JoinType joinType,
+            boolean maySkipOutputDuplicates,
+            Optional<DistributionType> distributionType,
+            Optional<Boolean> spillable,
+            List<String> dynamicFilterIds,
+            Optional<PlanNodeStatsAndCostSummary> reorderJoinStatsAndCost,
+            Map<AttributeKey, Object> leftAttributes,
+            Map<AttributeKey, Object> rightAttributes)
+    {
+        super(TRINO, NAME);
+        requireNonNull(resultName, "resultName is null");
+        requireNonNull(left, "left is null");
+        requireNonNull(right, "right is null");
+        requireNonNull(leftCriteriaSelector, "leftCriteriaSelector is null");
+        requireNonNull(rightCriteriaSelector, "rightCriteriaSelector is null");
+        requireNonNull(filter, "filter is null");
+        requireNonNull(leftHashSelector, "leftHashSelector is null");
+        requireNonNull(rightHashSelector, "rightHashSelector is null");
+        requireNonNull(leftOutputSelector, "leftOutputSelector is null");
+        requireNonNull(rightOutputSelector, "rightOutputSelector is null");
+        requireNonNull(dynamicFilterTargetSelector, "dynamicFilterTargetSelector is null");
+        requireNonNull(joinType, "joinType is null");
+        requireNonNull(distributionType, "distributionType is null");
+        requireNonNull(spillable, "spillable is null");
+        requireNonNull(dynamicFilterIds, "dynamicFilterIds is null");
+        requireNonNull(dynamicFilterIds, "dynamicFilterIds is null");
+        requireNonNull(reorderJoinStatsAndCost, "reorderJoinStatsAndCost is null");
+        requireNonNull(leftAttributes, "leftAttributes is null");
+        requireNonNull(rightAttributes, "rightAttributes is null");
+
+        if (!IS_RELATION.test(trinoType(left.type())) || !IS_RELATION.test(trinoType(right.type()))) {
+            throw new TrinoException(IR_ERROR, "left and right sources of Join operation must be of relation type");
+        }
+        this.left = left;
+        this.right = right;
+
+        if (leftCriteriaSelector.parameters().size() != 1 ||
+                !trinoType(leftCriteriaSelector.parameters().getFirst().type()).equals(relationRowType(trinoType(left.type()))) ||
+                !(trinoType(leftCriteriaSelector.getReturnedType()) instanceof RowType || trinoType(leftCriteriaSelector.getReturnedType()).equals(EMPTY_ROW))) {
+            throw new TrinoException(IR_ERROR, "invalid left criteria selector for Join operation");
+        }
```

Comment:
There is a lot of repetition of this code across different operation classes. It will be extracted and reused.